armi/bookkeeping/db/database3.py

```diff
@@ -100,6 +94,168 @@ def getH5GroupName(cycle: int, timeNode: int, statePointName: str = None) -> str
     return "c{:0>2}n{:0>2}{}".format(cycle, timeNode, statePointName or "")
 
 
+class JaggedArray:
+    def __init__(self, jaggedData, paramName):
+        """
+        Take a list of numpy arrays or lists and flatten them into a single 1D array.
+
+        This implementation can preserve the structure of a multi-dimensional numpy array
+        by storing the dimensions in self.shapes and then re-populating a numpy array of
+        that shape from the flattened 1D array. However, it can only preserve one layer of
+        jaggedness in a list of lists (or other iterables). For example, a list of tuples
+        with varying lengths can be flattened and reconstituted exactly. But, if a list of
+        lists of tuples is passed in, the tuples in that final layer of nesting will all be
+        flattened to a single 1D numpy array after a round trip. No structure is retained
+        from nested lists of lists or lists of tuples.
+
+        Parameters
+        ----------
+        jaggedData: list of numpy.ndarray
+            A list of numpy arrays (or lists) to be flattened into a single array
+        paramName: str
+            The name of the parameter represented by this data
+        """
+        offset = 0
+        flattenedArray = []
+        offsets = []
+        shapes = []
+        for array in jaggedData:
+            offsets.append(offset)
+            if isinstance(array, numpy.ndarray):
+                shapes.append(array.shape)
+                offset += array.size
+                flattenedArray.extend(array.flatten())
+            elif isinstance(array, (int, float, list, tuple)):
+                try:
+                    numpyArray = numpy.array(array)
+                    shapes.append(numpyArray.shape)
+                    offset += numpyArray.size
+                    flattenedArray.extend(numpyArray.flatten())
+                except:  # noqa: E722
+                    # numpy might fail if it's jagged
+                    flattenedList = self.flatten(array)
+                    shapes.append(
+                        len(flattenedList),
+                    )
+                    offset += len(flattenedList)
+                    flattenedArray.extend(flattenedList)
+            elif array is None:
+                flattenedArray.append(None)
+                offset += 1
+        self.flattenedArray = numpy.array(flattenedArray)
+        self.offsets = numpy.array(offsets)
+        self.shapes = numpy.array(shapes)
+        self.nones = numpy.array(
+            [i for i, val in enumerate(flattenedArray) if val is None]
+        )
+        self.dtype = self.flattenedArray.dtype
+        self.paramName = paramName
+
+    def __iter__(self):
+        """Iterate over the unpacked list."""
+        return iter(self.unpack())
+
+    def __contains__(self, other):
+        return other in self.flattenedArray
+
+    @staticmethod
+    def flatten(x):
+        if isinstance(x, list):
```

Comment:
I'm a little confused about this. Surely you could also flatting a tuple as well as a list?

If this is something to do with how you are treating tuples and lists differently, maybe:

1. We need a docstring explaining the limits of this method.
2. This should be a private method, not static, so no one outside this class can see it.

Thoughts?

Reply:
Yeah, it should include `tuple` and also probably `numpy.ndarray`, just in case.

I should add a docstring either way. The only reason I wouldn't want to make it private is so that the `JaggedArray` "quacks" like a numpy array with the `flatten()` method.